### PR TITLE
Add support for generating native libraries for Linux and macOS aarch64 platforms (fixes #46).

### LINKED ISSUE #46: CLP FFI for Linux/aarch64
### Request

I am getting the following exception while trying to build my application on `linux/aarch64`

```
2024-06-13T11:18:17.715+00:00 clp-ffi: Failed to find native library in JAR.
2024-06-13T11:18:17.717+00:00 clp-ffi: System.loadLibrary failed - no clp-ffi-java in java.library.path: /usr/java/packages/lib:/usr/lib64:/lib64:/lib:/usr/lib
```

Does CLP-FFI have a distribution for `linux/aarch64`?

### Possible implementation

N/A

### PR DESCRIPTION
**Pull request by kirkrodrigues**
# Description

This PR will allow users of aarch64 (a.k.a. arm64) platforms like Apple Silicon and ARM processors to use the Java clp-ffi library.

To achieve this, this PR:

* Adds a GH workflow job (`build-lib-for-linux-aarch64`) that uses Docker + QEMU to build the native library for aarch64 and upload it to GH Actions storage.
  * To run the build-env setup and compilation steps in the container, this PR also adds a shell script (`install-deps-and-build-native-lib.sh`).
* Updates the `build-lib-for-macos` GH workflow job to also build the native library for Apple Silicon using a [macos-14](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners/about-github-hosted-runners#standard-github-hosted-runners-for-public-repositories) GH-hosted runner, and upload the new binary to GH Actions storage.
  * This PR also removes the native-lib-build GH action since it was only being used on macOS and it's simple-enough to inline.
* Updates the `build-and-release` GH workflow job to download the new binaries and add them to the JAR.
* Modifies the `build-and-release` GH workflow job to build and test, add the binaries to the package, and then build the final JAR, rather than unnecessarily building a JAR before adding the additional binaries.
* Adds a GH job step to upload the final JAR to GH Actions storage for debugging.
* Adds an integration test which is an independent Maven project that uses the built JAR to run a simple program that checks whether the native library can be loaded and used correctly.
  * This PR also adds an `integration-tests` task to run this test, renames the current `test` to `unit-tests`, and adds a new `test` task to run `unit-tests` and `integration-tests`.

# Validation performed

* Downloaded the JAR built by the GH workflow.
* Ran `task integration-tests` to set up and run the integration tests.
* Copied the downloaded JAR to to `target/repo/com/yscope/clp/clp-ffi/0.4-SNAPSHOT`
* Validated that running the integration test manually (without using `task integration-tests`, since that would overwrite the downloaded JAR) passed on:
  * Ubuntu 20.04 (Focal) on the following platforms:
    * x86_64
    * aarch64
  * Ubuntu 22.04 (Jammy) on the following platforms:
    * x86_64
    * aarch64
  * macOS 14.2.1 on the following platforms:
    * aarch64

### CODE REVIEW

**Comment by kirkrodrigues** [comment]
Verified with @LinZhihao-723's help that the built jar works on an Apple Silicon Mac.

**Review by LinZhihao-723 — APPROVED** [review]

**Comment by davemarco** [comment]
Tested integration test on AWS x86 and arm64 22.04 ubuntu instances with this script. Printed output from test as expected. Did not test mac os as expensive on AWS. Maybe for future PR add test directly in workflow but not required for now. 

````
# Ubuntu setup
sudo apt update
sudo apt install -y build-essential cmake git maven openjdk-11-jdk
# Install task using your preferred method

# Testing steps
git clone git@github.com:y-scope/clp-ffi-java.git
cd clp-ffi-java
git submodule update --init --recursive
task integration-tests
cd target/repo/com/yscope/clp/clp-ffi/0.4-SNAPSHOT/
# Copy the .jar files into the dir above
cd ../../../../../../../integration-tests/jar-load-native-lib
MAVEN_OPTS="-Dmaven.repo.local=$(readlink -f ../../target/repo)" mvn compile
MAVEN_OPTS="-Dmaven.repo.local=$(readlink -f ../../target/repo)" mvn exec:java -Dexec.mainClass="com.yscope.clp.Main"
````

File: tools/scripts/ubuntu-focal/install-deps-build-test-package-native-lib.sh [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+set -e
+
+script_dir="$( cd "$( dirname "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )"
+project_dir="${script_dir}/../../../"
+
+apt-get update
+DEBIAN_FRONTEND=noninteractive apt-get install -y build-essential cmake maven
+
+cd "$project_dir"
+mvn --batch-mode validate generate-resources test assembly:single@assemble-lib-dir
```
Comment (kirkrodrigues) [inline tools/scripts/ubuntu-focal/install-deps-build-test-package-native-lib.sh:13]: I moved `validate` before `generate-resources` to match the order that Maven executes them in (see `mvn buildplan:list`).

**Comment by davemarco** [comment]
Tested again integration test had no errors after new changes. Checks all passes

**Review by davemarco — APPROVED** [review]

**Pull request merged**